Find newer versions of LIBXSMM and add clang-11 to CI

nilsdeppe opened the pull request:
## Proposed changes

- LIBXSMM has changed where they store version info yet again, so add another file to search. There is a major bug in LIBXSMM 1.9 that's fixed at least in 1.16.1 (I'm not sure exactly where it got fixed) but LIBXSMM 1.9 (what we are using in CI) fails to multiply 1x2 matrices (maybe other single-row matrices, haven't tested). Before updating to and requiring 1.16.1 I want to add the ability to correctly identify the newer versions so CI isn't broken.
- Add clang-11 to container and to CI
- I've added the LIBXSMM 1.16.1 switch in the Dockerfile here so I can just push the new container once this is merged. I'll submit a separate PR to mark the version as required.

I tried add GCC 10 but get the following failure:
```
2020-12-23T02:39:33.3069660Z           Start 1124: Unit.Utilities.PrettyType.short_name
2020-12-23T02:39:33.6569809Z 1073/1142 Test #1124: Unit.Utilities.PrettyType.short_name ........................................***Failed    0.35 sec
2020-12-23T02:39:33.6570416Z 
2020-12-23T02:39:33.6570762Z ############ ASSERT FAILED ############
2020-12-23T02:39:33.6571131Z Node: 0 Proc: 0
2020-12-23T02:39:33.6571627Z Line: 22 of /__w/spectre/spectre/src/Utilities/PrettyType.cpp
2020-12-23T02:39:33.6572857Z 'next_target != remainder->npos' violated!
2020-12-23T02:39:33.6573863Z Function: void pretty_type::detail::{anonymous}::remove_until(gsl::not_null<std::basic_string_view<char>*>, const char*)
2020-12-23T02:39:33.6574562Z Overran string DnEEE looking for 0123456789LNS
2020-12-23T02:39:33.6574980Z ############ ASSERT FAILED ############
2020-12-23T02:39:33.6575222Z 
2020-12-23T02:39:33.6575711Z Charm++> No provisioning arguments specified. Running with a single PE.
2020-12-23T02:39:33.6576746Z          Use +auto-provision to fully subscribe resources or +p1 to silence this message.
2020-12-23T02:39:33.6577408Z Charm++: standalone mode (not using charmrun)
2020-12-23T02:39:33.6577952Z Charm++> Running in Multicore mode: 1 threads (PEs)
2020-12-23T02:39:33.6578633Z Converse/Charm++ Commit ID: v6.10.2-0-g7bf00fa
2020-12-23T02:39:33.6580196Z Warning> Randomization of virtual memory (ASLR) is turned on in the kernel, thread migration may not work! Run 'echo 0 > /proc/sys/kernel/randomize_va_space' as root to disable it, or try running with '+isomalloc_sync'.
2020-12-23T02:39:33.6581202Z CharmLB> Load balancer assumes all CPUs are same.
2020-12-23T02:39:33.6581963Z Charm++> Running on 1 hosts (1 sockets x 2 cores x 1 PUs = 2-way SMP)
2020-12-23T02:39:33.6582536Z Charm++> cpu topology info is gathered in 0.000 seconds.
2020-12-23T02:39:33.6583043Z SpECTRE Build Information:
2020-12-23T02:39:33.6583428Z Version:                      2020.12.7
2020-12-23T02:39:33.6584175Z Compiled on host:             63db91c3ad25
2020-12-23T02:39:33.6584734Z Compiled in directory:        /work/build
2020-12-23T02:39:33.6585348Z Source directory is:          /__w/spectre/spectre
2020-12-23T02:39:33.6585961Z Compiled on git branch:       container_updates
2020-12-23T02:39:33.6586572Z Compiled on git revision:     NOT_IN_GIT_REPO
2020-12-23T02:39:33.6587260Z Linked on:                    Wed Dec 23 02:35:16 2020
2020-12-23T02:39:33.6588032Z Filters: "Unit.Utilities.PrettyType.short_name"
2020-12-23T02:39:33.6588588Z 
```
@wthrowe maybe you have some insight into this? I can file an issue if that's helpful

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Try to find LIBXSMM version yet another way

LIBXSMM seems to be unable to decide where it should stick the version number.

nilsdeppe committed:
Add clang-11 to container

nilsdeppe committed:
Add clang-11 and gcc-10 to CI

nilsdeppe committed:
Update Dockerfile to LIBXSMM 1.16.1

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -15,7 +15,7 @@
 FROM ubuntu:20.04
 
-ARG PARALLEL_MAKE_ARG=-j2
+ARG PARALLEL_MAKE_ARG=-j60
```
Let's not do that 😄 (you can pass `--build-arg PARALLEL_MAKE_ARG=-j60` to `docker build`)

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -47,6 +47,11 @@ RUN apt-get update -y \
                           libboost-thread-dev libboost-tools-dev libssl-dev \
                           libbenchmark-dev
 
+# Install clang-11
```
I just noticed this about the Charm++ build in the container, could you take a look while you're updating the container:

https://github.com/sxs-collaboration/spectre/blob/develop/containers/Dockerfile.buildenv#L217
```sh
# We do not build  with debug symbols and build with O2 to reduce build size.
```

This comment seems wrong, since the compiler flags are `-g -O0`. Could you update the comment or the flags?

wthrowe commented:
I can reproduce the demangler problem locally.  Looks like gcc changed its mangling of nullptr to correctly implement the standard.  I'll get a fix in.

kidder reviewed (commented):
I have no other requested changes

wthrowe commented:
I've submitted a fix for the demangler issue as #2709.

nilsdeppe committed:
Build charm++ with O2 in container

nilsdeppe committed:
Work around segfault in GCC-10 rel in Test_BulgedCube

Making one_over_rho_eta_zeta a ReturnType object instead of an expression
works around a weird issue in GCC-10 release mode (either a bug in the
optimizer or an unusual edge-case in Blaze) where the expressions in
radial_scaling_vector somehow have their internal pointers that should
point to cap_xi, cap_eta, and cap_zeta invalidated. Since the pointer
variable is optimized out it's unclear whether the pointer is set to
nullptr or something else goes wrong.

nilsdeppe commented:
I've rebased this on develop. Here are a few notes:

- GCC-10 debug is enabled in CI. There are segfaults in the following tests in GCC-10 release:
  - Unit.Domain.CoordinateMaps.BulgedCube (fixed in this PR)
  - Unit.Evolution.Systems.Cce.Equations
  - Unit.Evolution.Systems.Cce.InitializeJ
  - Unit.Evolution.Systems.Cce.NewmanPenrose

  It's unclear whether the issue is a GCC-10 optimizer bug or a Blaze bug. I'd say we wait until Blaze is upgraded to 3.9 to see if the issues persist and then make code changes as necessary. There's no clear way of handling this other than to break the math expressions over several lines.
- Changed Charm++ to be built with `-O2`
- clang-tidy failures are Boost TOMS748 root finders in BulgedCube

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -225,9 +225,9 @@ RUN git clone --single-branch --branch v6.10.2 --depth 1 \
     && git checkout v6.10.2 \
     && ./build LIBS multicore-linux-x86_64 gcc \
-      ${PARALLEL_MAKE_ARG} -g -O0 --build-shared \
+      ${PARALLEL_MAKE_ARG} -g -O2 --build-shared \
```
The comment says "we do not build with debug symbols", so should this `-g` be deleted?

nilsdeppe replied:
Ah, yea. This is a back and forth thing. If we need to debug something with `pup` it is potentially useful to have the debug symbols. I'm not sure which is the right thing to do, to be honest. Basically it's:
1. smaller container by not have debug symbols (not sure how much smaller)
2. have debug symbols to make debugging Charm++-related issues easier for folks using the container

nilsvu replied:
I think we should make the container debug-friendly, since it's our dev and debug environment. That's to a reasonable extent of course, so if debug symbols for charm were unreasonably large we could consider omitting them, but that's probably not the case. If we end up running production simulations in the container we should make a production container anyway, which would omit a lot of debug/CI things like the older compiler versions.

nilsvu replied:
@nilsdeppe do you want to make another change to the Dockerfile or should we merge as-is, potentially making further changes in follow-up PRs?

nilsdeppe replied:
Rebased and squashed in updated comment :)

nilsvu reviewed (dismissed):
LGTM, I'm fine merging as-is and fixing the Charm installation in the container later, if there's anything to fix.

fmahebert commented on .github/workflows/Tests.yaml:
```diff
@@ -339,6 +341,16 @@ jobs:
+          # - Unit.Evolution.Systems.Cce.InitializeJ
+          # - Unit.Evolution.Systems.Cce.NewmanPenrose
+          # Test_BulgedCube is fixed. This is either a GCC-10 optimizer bug or
+          # some weird Blaze behavior with GCC-10.
```
[opinion, you may choose to disagree] Delete this last sentence, which isn't useful to a future reader of the comment

nilsdeppe replied:
I disagree because I think it is useful to forward as much information as I can about what the underlying bug may be. It's not a lot (or really any) information, but I think it's better than none :shrug: I think knowing that the issue is related to Blaze math expressions is useful since I do know that this is why the segfaults are occurring. I was unable to figure out the underlying Blaze & compiler interaction that resulted in the segfault, though.

nilsvu commented:
ignoring clang-tidy complaints from boost root-finding
